# Borrowers reliability research


## Let's open the table and study the general information about the data.

**Task 1. Import the pandas library. Read data from a CSV file into a DataFrame and save it to the variable data. Provide the file path:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Task 2. Display the first 20 rows of the DataFrame data on the screen.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Task 3. Display basic information about the DataFrame using the info() method.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Data preprocessing

### Removing missing values

**Task 4. Display the number of missing values for each column. Use a combination of two methods.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Task 5. There are missing values in two columns. One of them is days_employed. The missing values in this column will be handled in the next step. The other column with missing values is total_income, which contains income data. The type of employment has the most significant impact on income, so the missing values in this column should be filled with the median value for each type in the income_type column. For example, if a person has an employment type of employee, the missing value in the total_income column should be filled with the median income among all records with the same employment type.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Handling anomalous values

**Task 6. The data may contain artifacts (anomalies) – values that do not reflect reality and appeared due to some error. An example of such an artifact is the negative number of days of employment in the days_employed column. For real data, this is normal. Handle the values in this column: replace all negative values with positive ones using the abs() method.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Task 7. For each type of employment, display the median value of employment duration (days_employed) in days**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (unemployed and retirees) will have anomalously large values. Correcting such values is difficult, so we will leave them as they are.

**Task 8. Display the list of unique values in the children column.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Task 9. There are two anomalous values in the children column. Remove the rows containing these anomalous values from the DataFrame data.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Task 10. Display the list of unique values in the children column again to ensure that the artifacts have been removed.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Dropping missing values

**Task 11. Fill missing values in the days_employed column with the median values grouped by each income_type.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Task 12. Ensure that all missing values are filled. Double-check and display the number of missing values for each column using two methods.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Changing data types

**Task 13. Convert the floating-point data type in the total_income column to integer using the astype() method.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Handling duplicates

**Task 14. Handle implicit duplicates in the education column. There are identical values written in different cases (uppercase and lowercase). Convert them to lowercase.**

In [14]:
data['education'] = data['education'].str.lower()

**Task 15. Display the number of duplicate rows in the data. If such rows exist, remove them..**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Data categorization

**Task 16. Based on the ranges provided below, create a column total_income_category in the DataFrame data with categories:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**For example, assign category 'E' to a borrower with an income of 25,000 and category 'B' to a client earning 235,000. We'll use a custom function named categorize_income() along with the apply() method.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Task 17. Display the list of unique purposes for taking a loan from the purpose column.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Task 18. Create a function that, based on the data in the purpose column, will form a new column purpose_category with the following categories:

'car operations',
'real estate operations',
'wedding ceremony',
'education'.
For example, if the purpose column contains the substring 'car purchase', the purpose_category column should contain 'car operations'.

Use a custom function named categorize_purpose() along with the apply() method. Review the data in the purpose column to determine which substrings will help correctly categorize them..**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Step 3. Let's explore the data and answer questions

#### 3.1 Is there a correlation between the number of children and timely loan repayment?

Let's call a function that will calculate the proportion of borrowers with overdue payments for each client category based on the number of children.

In [ ]:
def my_mean(x):
    return '{:.2%} '.format(x.mean())
# Вызовем функцию
data.groupby('children')['debt'].agg(['sum','count', my_mean])

sum  count my_mean
children                     
0         1063  14091  7.54% 
1          444   4808  9.23% 
2          194   2052  9.45% 
3           27    330  8.18% 
4            4     41  9.76% 
5            0      9  0.00%

Due to the small number of borrowers with 3+ children, we'll categorize and distribute borrowers as follows:

Borrowers with no children fall into category "0".
Borrowers with 1 child fall into category "1".
Borrowers with 2 or more children belong to category "2+".

In [23]:
def categorize_children(children):
    if children == 0:
        return 0
    elif children == 1:
        return 1
    else:
        return '2+'

data['children_group'] = data['children'].apply(categorize_children)
result = data.groupby('children_group')['debt'].agg(['sum','count', my_mean])
print(result)

                 sum  count my_mean
children_group                     
0               1063  14091  7.54% 
1                444   4808  9.23% 
2+               225   2432  9.25% 


**Conclusion**
Based on the results obtained, it can be concluded that there is a correlation between the number of children and timely loan repayment. Families without children have fewer additional expenses and tend to service their loans on time. Clients with children face higher family budget expenditures, which can lead to difficulties in making timely payments on credit obligations.

#### 3.2 Is there a correlation between marital status and timely loan repayment?

In [24]:
data.groupby(['family_status','family_status_id'])['debt'].agg(['sum','count', my_mean]).sort_values('family_status_id')

sum  count my_mean
family_status         family_status_id                    
женат / замужем       0                 927  12261  7.56% 
гражданский брак      1                 385   4134  9.31% 
вдовец / вдова        2                  63    951  6.62% 
в разводе             3                  84   1189  7.06% 
Не женат / не замужем 4                 273   2796  9.76%

**Вывод:** 
Based on the results obtained, unmarried individuals and those in civil partnerships have the worst loan servicing records.

#### 3.3 Is there a correlation between income level and timely loan repayment?

In [25]:
data.groupby(['total_income_category'])['debt'].agg(['sum','count', my_mean]).sort_values('my_mean')

sum  count my_mean
total_income_category                     
D                        21    349  6.02% 
B                       354   5014  7.06% 
A                         2     25  8.00% 
C                      1353  15921  8.50% 
E                         2     22  9.09%

#### 3.4 How do different loan purposes affect timely repayment?

In [26]:
data.groupby(['purpose_category'])['debt'].agg(['sum','count', my_mean]).sort_values('my_mean')

sum  count my_mean
purpose_category                            
операции с недвижимостью  780  10751  7.26% 
проведение свадьбы        183   2313  7.91% 
получение образования     369   3988  9.25% 
операции с автомобилем    400   4279  9.35%

**ВConclusion:** 
- Clients who took out loans for real estate operations or wedding ceremonies tend to have better repayment records.
- Clients who took out loans for car operations or education tend to have poorer repayment records.


#### 3.5 Possible reasons for missing values in the original data could include:

*Answer:* 
- Human error: data entry mistakes, reluctance of clients to answer specific questions.
- Technical issues: equipment or software malfunctions during data collection. Missing data could be removed during data processing or cleaning.

#### 3.6 Explain why filling missing values with the median is the best solution for quantitative variables.

*Answer:* 
Extreme values - excessively large/small values can distort calculation results.
The median is less sensitive to outliers compared to the mean.

### Step 4: overall conclusion.

During the investigation of borrower reliability, the following actions were taken:

The received data was reviewed.
Missing values were processed and filled.
Duplicates were removed.
Clients were grouped based on various parameters.
Upon examining the first 20 rows of the file, it became evident that the data contains incorrect and illogical values (anomalies). All of these anomalies significantly interfere with data processing and can heavily distort results.

For example:

The days_employed column, describing the employment length of borrowers, contains negative values.
The education column, reflecting the educational level of borrowers, has words with letters written in chaotic cases.
The purpose column, containing the reason for taking the loan, includes strings with the same meaning but different formulations.
Based on the analysis conducted to address the questions posed, the following conclusions can be drawn:

**Clients who are married, have no children, and take out loans for real estate-related purposes exhibit the best loan servicing records.**:

- Borrowers with the status "Single / Never Married" have the highest loan repayment default rate at 9.76%.
- Borrowers with the status "Widowed" have the lowest loan repayment default rate at 6.62%. Widows and widowers are typically older individuals who tend to be more responsible. 

In terms of the "income level/loan servicing quality" relationship, the following observations can be made:

The highest proportion of clients with overdue debt is concentrated in category B, indicating that clients with lower incomes are less diligent in servicing their credit obligations.

Clients who took out loans for real estate operations exhibit the best repayment rates (7.26% default rate).

Clients who took out loans for car operations and education exhibit the poorest repayment rates (9.35% and 9.25% default rates, respectively).

